org.eclipse.scanning.api/src/org/eclipse/scanning/api/scan/event/Location.java

```diff
@@ -1,150 +1,150 @@
+
+import java.util.ArrayList;
+import java.util.List;
+
+import org.eclipse.scanning.api.ILevel;
+import org.eclipse.scanning.api.INameable;
+import org.eclipse.scanning.api.points.IPosition;
+import org.eclipse.scanning.api.scan.PositionEvent;
+
+/**
+ *
+ * A location is a json object for notifying over JMS
+ * the value of an IScannable. We do not use IPosition
+ * directly because this does not have all the information
+ * required. A location may be a transient state which a
+ * motor is moving through or a notifcation of an intent
+ * to move. It can be thought of as a PositionEvent
+ * encapsulated in a single JSON object and with context
+ * information.
+ *
+ * @author Matthew Gerring
+ *
+ */
+public class Location implements INameable {
+
+	/**
+	 * Must match method names in IPositionListener as reflection used on the value to
+	 * call the correct method on the listener.
+	 */
+	public enum LocationType {
+		positionWillPerform, levelPerformed, positionChanged, positionPerformed;
+	}
+
+
+	private LocationType type;
+	private String       name;
+	private IPosition    position;
+	private int          level;
+	private List<String> levelNames;
+
+	public Location() {
+
+	}
+	public Location(LocationType t, PositionEvent evnt) {
+		this.type       = t;
+		this.position   = evnt.getPosition();
+		if (position.getNames()!=null&&position.getNames().size()==1) this.name = position.getNames().iterator().next();
+		this.level      = evnt.getLevel();
+		this.levelNames = readLevelNames(evnt.getLevelObjects());
+	}
+
+	private List<String> readLevelNames(List<? extends ILevel> levelObjects) {
+
+		if (levelObjects==null || levelObjects.isEmpty()) return null;
+		List<String> ret = new ArrayList<>();
+		for (ILevel level : levelObjects) {
+			ret.add(level.getName());
+		}
+		return ret;
+	}
+	public LocationType getType() {
+		return type;
+	}
+	public void setType(LocationType type) {
+		this.type = type;
+	}
+	@Override
+	public String getName() {
+		return name;
+	}
+	@Override
+	public void setName(String name) {
+		this.name = name;
+	}
+	public IPosition getPosition() {
+		return position;
+	}
+	public void setPosition(IPosition position) {
+		this.position = position;
+	}
+	public int getLevel() {
+		return level;
+	}
+	public void setLevel(int level) {
+		this.level = level;
+	}
+	public List<String> getLevelNames() {
+		return levelNames;
+	}
+	public void setLevelNames(List<String> levelNames) {
+		this.levelNames = levelNames;
+	}
+	@Override
+	public int hashCode() {
+		final int prime = 31;
+		int result = 1;
+		result = prime * result + level;
+		result = prime * result + ((levelNames == null) ? 0 : levelNames.hashCode());
+		result = prime * result + ((name == null) ? 0 : name.hashCode());
+		result = prime * result + ((position == null) ? 0 : position.hashCode());
+		result = prime * result + ((type == null) ? 0 : type.hashCode());
+		return result;
+	}
+	@Override
+	public boolean equals(Object obj) {
```

Comment:
![CRITICAL](https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL') Refactor this method to reduce its Cognitive Complexity from 17 to the 15 allowed. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS3776)